protocols/gossipsub/src/handler.rs

```diff
@@ -272,49 +277,65 @@ where
+                                }
+                                Err(e) => {
+                                    return Poll::Ready(ProtocolsHandlerEvent::Close(e));
+                                }
+                            }
+                        }
+                        Poll::Ready(Err(e)) => {
+                            // TODO: Is this what we want?
+                            return Poll::Ready(ProtocolsHandlerEvent::Close(e));
                         }
-                        AsyncSink::NotReady(message) => {
+                        Poll::Pending => {
                             self.outbound_substream =
                                 Some(OutboundSubstreamState::PendingSend(substream, message));
                             break;
                         }
                     }
                 }
                 Some(OutboundSubstreamState::PendingFlush(mut substream)) => {
-                    match substream.poll_complete()? {
-                        Async::Ready(()) => {
+                    match Sink::poll_flush(Pin::new(&mut substream), cx) {
+                        Poll::Ready(Ok(())) => {
                             self.outbound_substream =
                                 Some(OutboundSubstreamState::WaitingOutput(substream))
                         }
-                        Async::NotReady => {
+                        Poll::Ready(Err(e)) => {
+                            // TODO: Is this what we want?
+                            return Poll::Ready(ProtocolsHandlerEvent::Close(e))
+                        }
+                        Poll::Pending => {
                             self.outbound_substream =
                                 Some(OutboundSubstreamState::PendingFlush(substream));
                             break;
                         }
                     }
                 }
                 // Currently never used - manual shutdown may implement this in the future
-                Some(OutboundSubstreamState::_Closing(mut substream)) => match substream.close() {
-                    Ok(Async::Ready(())) => {
+                Some(OutboundSubstreamState::_Closing(mut substream)) => match Sink::poll_close(Pin::new(&mut substream), cx) {
+                    Poll::Ready(Ok(()) )=> {
                         self.outbound_substream = None;
                         if self.inbound_substream.is_none() {
                             self.keep_alive = KeepAlive::No;
                         }
                         break;
                     }
-                    Ok(Async::NotReady) => {
-                        self.outbound_substream = Some(OutboundSubstreamState::_Closing(substream));
-                        break;
-                    }
-                    Err(_) => {
-                        return Err(io::Error::new(
+                    Poll::Ready(Err(_)) => {
+                        // TODO: Why do we not at least log the error (`_`)?
```

Comment:
@AgeManning what do you think?

Reply:
Yeah I think we should log the error. I think the first time I wrote this I was trying to match the go-impl so was in a hurry.